velox/expression/EvalCtx.h

```diff
@@ -27,8 +27,139 @@ namespace facebook::velox::exec {
+      bool canPeelsHaveNulls);
+
+  /// Utility method used to check whether an encoding is peel-able.
+  inline static bool isPeelable(VectorEncoding::Simple encoding) {
+    switch (encoding) {
+      case VectorEncoding::Simple::CONSTANT:
+      case VectorEncoding::Simple::DICTIONARY:
+      case VectorEncoding::Simple::SEQUENCE:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  /// Return the encoding of the peeled wrap.
+  /// Should only be called if getPeeledVectors() was successful.
+  VectorEncoding::Simple getWrapEncoding() const;
+
+  /// Translates row numbers of the outer vector via the peel into row numbers
+  /// of the inner vector.
+  /// Should only be called if getPeeledVectors() was successful.
+  SelectivityVector* translateToInnerRows(
+      const SelectivityVector& rows,
+      LocalSelectivityVector& newRowsHolder);
+
+  /// Wraps the given vector 'peeledResult' with the peeled encoding and
+  /// generates a vector which has valid rows as per the input 'rows'
+  /// selectivity vector.
+  /// Should only be called if getPeeledVectors() was successful.
+  VectorPtr wrap(
+      const TypePtr& outputType,
+      velox::memory::MemoryPool* FOLLY_NONNULL pool,
+      VectorPtr peeledResult,
+      const SelectivityVector& rows);
+
+  /// Takes a set of outer rows (rows defined over the peel) and a functor that
+  /// is executed for each outer row which does not map to a null row.
+  /// The outer and its corresponding inner row is passed to the functor.
+  void applyToNonNullInnerRows(
+      const SelectivityVector& outerRows,
+      std::function<void(vector_size_t, vector_size_t)> func) {
+    auto indices = wrap_ ? wrap_->as<vector_size_t>() : nullptr;
+    auto wrapNulls = wrapNulls_ ? wrapNulls_->as<uint64_t>() : nullptr;
+    outerRows.applyToSelected([&](auto outerRow) {
+      // A known null in the outer row masks an error.
+      if (wrapNulls && bits::isBitNull(wrapNulls, outerRow)) {
+        return;
+      }
+      vector_size_t innerRow = indices ? indices[outerRow] : constantWrapIndex_;
+      func(outerRow, innerRow);
+    });
+  };
+
+ private:
+  void setDictionaryWrapping(
+      DecodedVector& decoded,
+      const SelectivityVector& rows,
+      BaseVector& firstWrapper);
+
+  /// The encoding of the peel. Set after getPeeledVectors() is called. Is equal
+  /// to Flat if getPeeledVectors() has not been called or peeling was not
+  /// successful.
+  VectorEncoding::Simple wrapEncoding_ = VectorEncoding::Simple::FLAT;
+  /// The dictionary indices. Only valid if wrapEncoding_ = DICTIONARY.
+  BufferPtr wrap_;
+  /// The dictionary nulls. Only valid if wrapEncoding_ = DICTIONARY.
+  BufferPtr wrapNulls_;
+  /// The size of one of the peeled vectors. Only valid if wrapEncoding_ =
+  /// DICTIONARY.
+  vector_size_t baseSize_ = 0;
+  /// The constant index. Only valid if wrapEncoding_ = CONSTANT.
+  vector_size_t constantWrapIndex_ = 0;
+};
+
```

Comment:
I got spoiled with the refactor and was looking for a PeeledEncodingTest.cpp file somewhere. Would it make sense to create a test suite to test this API directly?